### PR TITLE
Decouple Player and PlayerCommands a little bit more

### PR DESCRIPTION
**Pull request by policiu**
Justifications:

Player -> IMoveableEntity

We are trying to separate our PlayerCommands from the Player. If we implement an interface, we can more easily reuse Dash. For example using it for an enemy, different Player base.  

Input to its own Class/struct:

Same reason as before, the less reliance on player the better.

Timers as a property:

Timer relies on a Node, which the commands are not

Velocity field removal:
It was a tad confusion and was modifying a struct, now functions just update the velocity directly.

### CODE REVIEW

**Comment by policiu** [comment]
Can't really do too much more without knowing how wea re using signals/events/observer.

File: Scripts/Player/PlayerCommand.cs [hunk 1]
```diff
@@ -2,18 +2,90 @@
 
 namespace Sankari;
 
+public interface IMoveableEntity
+{
+	// Velocity of the entity
+	Vector2 Velocity { get; set; }
+
+	// Sprite to modify
+	AnimatedSprite2D AnimatedSprite { get; }
```
Comment (valkyrienyanko) [inline Scripts/Player/PlayerCommand.cs:9]: Not every moving entity might be a animated sprite, could be just a sprite instead of a animated sprite, or in a very unlikely scenario a ColorRect node.

File: Scripts/Player/PlayerCommand.cs [hunk 2]
```diff
@@ -2,18 +2,90 @@
...
+
+	// Max speed on the ground while sprinting
+	int SpeedMaxGroundSprint { get; }
```
Comment (valkyrienyanko) [inline Scripts/Player/PlayerCommand.cs:12]: What if this moving entity only flys and has nothing to do with the ground?

File: Scripts/Player/PlayerCommand.cs [hunk 3]
```diff
@@ -2,18 +2,90 @@
...
+	int SpeedMaxAir { get; }
+
+	// Current direction the entity is moving
+	Vector2 MoveDir { get; }
+
+	// Position in the world
+	Vector2 GlobalPosition { get; }
+
+	// Speed given when dashing vertically
+	int SpeedDashVertical { get; }
```
Comment (valkyrienyanko) [inline Scripts/Player/PlayerCommand.cs:27]: Every moving entity has the ability to dash?

File: Scripts/Player/Player.cs [hunk 4]
```diff
@@ -86,6 +80,7 @@ public override void _Ready()
 		ParentWallChecksRight = GetNode<Node2D>(NodePathRayCast2DWallChecksRight);
 		AnimatedSprite        = GetNode<AnimatedSprite2D>("AnimatedSprite2D");
 		DieTween              = new GTween(this);
+		Timers                = new GTimers(this);
```
Comment (valkyrienyanko) [inline Scripts/Player/Player.cs:81]: I'm not sure I like what you've done here. I don't mind typing `this` for each timer that is created, I'm never going to remember that there is a `GTimers.cs` script when dealing with timers. This player script just happens to have a lot of timers, this isn't common across scripts.

File: Scripts/Player/PlayerCommand.cs [hunk 5]
```diff
@@ -2,18 +2,90 @@
 
 namespace Sankari;
 
+public interface IMoveableEntity
```
Comment (valkyrienyanko) [inline Scripts/Player/PlayerCommand.cs:3]: This should go in a file of its own under a folder called Models or Interfaces or something similar.

File: Scripts/Player/PlayerCommand.cs [hunk 6]
```diff
@@ -2,18 +2,90 @@
...
+	bool IsOnGround();
+
+	// Force the entity to jump
+	void Jump();
+}
+
 public abstract class PlayerCommand
 {
-	public virtual void Init(Player player)
+	protected IMoveableEntity Entity { get; set; }
```
Comment (valkyrienyanko) [inline Scripts/Player/PlayerCommand.cs:63]: Why not use `Player Player { get; set; }` here? I can't imagine what benefit passing a slime through PlayerCommand would have. Slime is a moveable entity after all.

**Review by valkyrienyanko — COMMENTED** [review]
Submitted review

**Comment by valkyrienyanko** [comment]
> Can't really do too much more without knowing how wea re using signals/events/observer.

Since all I know how to use is the `Notifications.cs` script, all I can recommend is that. The script could be left as is or it could be refactored to be split up a bit as everything is currently in one large dictionary. This could become slow? Perhaps if the dictionary was split into a dictionary per entity? or per scene? Or maybe this is just overkill and everything should be left as is. This is just me speaking my thoughts out loud. That is assuming you want to use this script.

File: Scripts/Player/PlayerCommand.cs (hunk 2 above)
Comment (policiu) [reply Scripts/Player/PlayerCommand.cs:12]: Good idea, I'll create interfaces for each

File: Scripts/Player/PlayerCommand.cs (hunk 6 above)
Comment (policiu) [reply Scripts/Player/PlayerCommand.cs:63]: The idea is to eventually remove "PlayerCommandDash" with "CommandDash".

Comment (valkyrienyanko) [reply Scripts/Player/PlayerCommand.cs:63]: What else would have a dash ability in the game?

File: Scripts/Player/Player.cs (hunk 4 above)
Comment (policiu) [reply Scripts/Player/Player.cs:81]: I'm a bit confused.
Why does the consumer (our commands) need to know that timers needs a Node? 

Why is it clearer and explicit? Is there a reason why putting this with a timer makes it clearer than asking the entity to start a timer?

File: Scripts/Player/PlayerCommand.cs (hunk 5 above)
Comment (policiu) [reply Scripts/Player/PlayerCommand.cs:3]: If each command has an interface, it might be nice to leave it as is. See my changes.

File: Scripts/Player/PlayerCommand.cs (hunk 3 above)
Comment (policiu) [reply Scripts/Player/PlayerCommand.cs:27]: Updated! Now each command has a set of requirements.

File: Scripts/Player/PlayerCommand.cs (hunk 1 above)
Comment (policiu) [reply Scripts/Player/PlayerCommand.cs:9]: This is a bit hard since the commands want it. I moved it out.

File: Sankari.csproj [hunk 7]
```diff
@@ -15,4 +15,7 @@
       <HintPath>ENet-CSharp.dll</HintPath>
     </Reference>
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Scripts\Commands\" />
```
Comment (valkyrienyanko) [inline Sankari.csproj:19]: Why is this needed here?

**Comment by valkyrienyanko** [comment]
> `public abstract class EntityCommand<T> : PlayerCommand`

Seeing EntityCommand<T> extend from PlayerCommand looks weird to me

**Comment by valkyrienyanko** [comment]
What is the point of making `IEntityDashable` interface if there will only ever be one entity (the player) that can dash?

File: Scripts/Player/Player.cs (hunk 4 above)
Comment (valkyrienyanko) [reply Scripts/Player/Player.cs:83]: I see what you're doing now, nvm, otherwise you would have to pass the node down the commands as a reference.

File: Scripts/Player/PlayerCommand.cs (hunk 5 above)
Comment (valkyrienyanko) [reply Scripts/Player/PlayerCommand.cs]: Not every command needs an interface if there will only ever be one entity using that command unless you're just trying to be consistent?

**Comment by policiu** [comment]
> > `public abstract class EntityCommand<T> : PlayerCommand`
> 
> Seeing EntityCommand extend from PlayerCommand looks weird to me

Whoops, fixed.


> What is the point of making `IEntityDashable` interface if there will only ever be one entity (the player) that can dash?
Better decoupling. Dash shouldn't care what its getting as long as it's something usable.

Also, There could be an enemy who hones in on the player with a dash
There could be modes where dashing isn't available and/or we use a different player class.
Maybe a replay ghost. It's pretty easy to decouple now before we do more things.

Regardless why does it need to be bound to the player class? I don't see a point in making it exclusive to the player.

File: Sankari.csproj (hunk 7 above)
Comment (policiu) [reply Sankari.csproj:19]: No idea, it was auto added when I made the folder in VS. Removed, probably fine.

File: Scripts/Player/PlayerCommand.cs (hunk 6 above)
Comment (policiu) [reply Scripts/Player/PlayerCommand.cs]: See comment below

File: Scripts/Commands/EntityCommandDash.cs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+	int SpeedDashVertical { get; }
+
+	// Speed given when dashing horizontally
+	int SpeedDashHorizontal { get; }
+	// Is the entity on the ground?
+
+	bool IsOnGround();
+	
+	// Sprite to modify
+	AnimatedSprite2D AnimatedSprite { get; }
```
Comment (valkyrienyanko) [inline Scripts/Commands/EntityCommandDash.cs:24]: Maybe remove this. Entities could be an AnimatedSprite, Sprite or ColorRect or maybe something else entirely.

File: Scripts/Player/Movement/MovementUtils.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+﻿namespace Sankari;
+
+public class MovementInput
```
Comment (valkyrienyanko) [inline Scripts/Player/Movement/MovementUtils.cs:3]: Perhaps this should be renamed to `PlayerInputMovement` and moved outside of "Movement" folder.

File: Scripts/Commands/EntityCommandDash.cs (hunk 8 above)
Comment (policiu) [reply Scripts/Commands/EntityCommandDash.cs:24]: Can't until we remove Sprite from both classes. You mentioned earlier about separating out sprites and movement commands, but that's a but out of scope for my changes.

I don't really thing there is anything too wrong with requiring a AnimatedSprite for now until we:
a. Find a way to separate this out (Signals/Events, or whatever)
b. Do something so that the commands don't need something that specific.

Regardless, I would ask when the separation happens please make it a pull request before merging (perhaps branch off of this shared branch for now.)

File: Scripts/Player/Movement/MovementUtils.cs (hunk 9 above)
Comment (policiu) [reply Scripts/Player/Movement/MovementUtils.cs]: Hmm, good point. I renamed the Utils function to mention Player, but we are going to want to figure out how we want to represent Movement Options for the Commands.

**Comment by valkyrienyanko** [comment]
Everything is looking good, should I merge this?

File: Scripts/Commands/EntityCommandDash.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+
+}
+
+public class EntityCommandDash : EntityCommand<IEntityDashable>
+{
+	public EntityCommandDash(IEntityDashable entity) : base(entity)
```
Comment (valkyrienyanko) [inline Scripts/Commands/EntityCommandDash.cs:31]: Would removing this line of code break anything as it is right now?

File: Scripts/Commands/EntityCommands.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+	/// <summary>
+	/// Called after most Movement logic occurs
+	/// </summary>
+	/// <param name="input">Input to act on</param>
+	public virtual void LateUpdate(MovementInput input)
+	{ }
+
+}
+
+public abstract class EntityCommand<T> : EntityCommand
```
Comment (valkyrienyanko) [inline Scripts/Commands/EntityCommands.cs:46]: Why not merge this generic <T> to class EntityCommand (line 22) so

`public abstract class EntityCommand`

becomes 

`public abstract class EntityCommand<T>`

instead of having both class versions. In other words is there a reason you separated them?

File: Scripts/Player/Player.cs [hunk 12]
```diff
@@ -1,6 +1,8 @@
 namespace Sankari;
 
-public partial class Player : CharacterBody2D
+interface IPlayerSkills : IEntityDashable, IEntityWallJumpable { }
```
Comment (valkyrienyanko) [inline Scripts/Player/Player.cs:3]: For consistency it wouldn't hurt to add `public` to start

File: Scripts/Commands/EntityCommands.cs [hunk 13]
```diff
@@ -0,0 +1,55 @@
+﻿namespace Sankari;
+
+public interface IEntityMoveable
+{
+	// Velocity of the entity
+	Vector2 Velocity { get; set; }
```
Comment (valkyrienyanko) [inline Scripts/Commands/EntityCommands.cs:6]: I'm assuming everything without an access modifier in an interface is public by default. Just not seeing public before each prop looks weird to me but I suppose I can get use to it.

File: Scripts/Utils/Godot Helpers/GTimers.cs [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+public class GTimers
+{
+	private Node Node { get; set; }
+
+	public GTimers(Node node)
+	{
+		Node = node;
+	}
+
+	/// <inheritdoc cref="GTimer.GTimer(Node, string, int, bool, bool)"/>
```
Comment (valkyrienyanko) [inline Scripts/Utils/Godot Helpers/GTimers.cs:12]: This XML looks foreign to me

File: Scripts/Utils/MovementUtils.cs [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+{
+	public bool IsJump { get; set; }
+	public bool IsUp { get; set; }
+	public bool IsDown { get; set; }
+	public bool IsFastFall { get; set; }
+	public bool IsDash { get; set; }
+	public bool IsSprint { get; set; }
+}
+
+public class MovementUtils
```
Comment (valkyrienyanko) [inline Scripts/Utils/MovementUtils.cs:13]: Both these classes should be moved back to Player folder as they are 100% related to Player and Player alone

File: Scripts/Player/Player.cs [hunk 16]
```diff
@@ -224,23 +245,25 @@ private async void CheckIfCanGoUnderPlatform(bool inputDown)
 		}
 	}
 
-	private void HorzDampening(int dampening)
+	private float HorzDampening(float number, uint dampening)
```
Comment (valkyrienyanko) [inline Scripts/Player/Player.cs:248]: What was your reasoning for these changes? (and all the changes you made below to this function?)

File: Scripts/Player/Player.cs [hunk 17]
```diff
@@ -209,13 +204,39 @@ public void Jump()
...
+	{
+		var left = IsTouchingWallLeft();
+		var right = IsTouchingWallRight();
+
+		return -Convert.ToInt32(left) + Convert.ToInt32(right);
+	}
+
+	private bool IsTouchingWallLeft()
+	{
+		foreach (var raycast in RayCast2DWallChecksLeft)
```
Comment (valkyrienyanko) [inline Scripts/Player/Player.cs:217]: Idea: A new function could be made `public static bool IsTouching(this List<RayCast2D> raycasts);` placed in `Scripts/Utils/Extensions/ExtensionsCollections.cs`

File: Scripts/Player/Player.cs [hunk 18]
```diff
@@ -129,76 +128,72 @@ private void NetUpdate()
...
+		WallDir = UpdateWallDirection();
 
-		CheckIfCanGoUnderPlatform(InputDown);
+		UpdateUnderPlatform(input);
 
 		foreach (var command in PlayerCommands)
-			command.Update(this);
+			command.Update(input);
+
+		HandleGravityState(delta, input);
```
Comment (valkyrienyanko) [inline Scripts/Player/Player.cs:139]: Feels weird to have this here now there are commands, but I can't think of where else to place this.

**Review by valkyrienyanko — APPROVED** [review]

File: Scripts/Commands/EntityCommandDash.cs (hunk 10 above)
Comment (policiu) [reply Scripts/Commands/EntityCommandDash.cs:31]: What part of it? The constructor? Yeah, C# default constructor is new this() and we don't have that constructor in the base (we expect new base(Entity)), so the implicit constructor fails since it can't override new base().

File: Scripts/Commands/EntityCommands.cs (hunk 11 above)
Comment (policiu) [reply Scripts/Commands/EntityCommands.cs:46]: Yeah, in C# you can't have collections of Templated classes, so we need a base class to reference.

E.g. Without the base class, List<EntityCommand> fails since a type is expect for T. Using this allows use to reference EntityCommand in a collection while allowing children to reference and inherit EntityCommand<T>

File: Scripts/Player/Player.cs (hunk 12 above)
Comment (policiu) [reply Scripts/Player/Player.cs:3]: Agreed.

File: Scripts/Commands/EntityCommands.cs (hunk 13 above)
Comment (policiu) [reply Scripts/Commands/EntityCommands.cs:6]: It should have public.

File: Scripts/Utils/Godot Helpers/GTimers.cs (hunk 14 above)
Comment (policiu) [reply Scripts/Utils/Godot Helpers/GTimers.cs:12]: It basically says "use the XML from cref". Basically, I don't need to rewrite the XML since it'll use the GTimer.GTimer constructor XML.

File: Scripts/Utils/MovementUtils.cs (hunk 15 above)
Comment (policiu) [reply Scripts/Utils/MovementUtils.cs:13]: Sounds good.

File: Scripts/Player/Player.cs (hunk 16 above)
Comment (policiu) [reply Scripts/Player/Player.cs:248]: Ah, it was an artifact from my original changes. I can revert if wanted. (Basically I was trying to make the function not rely on the field playerVelocity and eventually remove playerVelocity which I did.)

File: Scripts/Player/Player.cs (hunk 18 above)
Comment (policiu) [reply Scripts/Player/Player.cs:139]: Same, it might be nice to change it in the future once we figure out what is happening with Sprites.

File: Scripts/Player/Player.cs [hunk 19]
```diff
@@ -103,6 +93,15 @@ public override void _Ready()
 
 		FloorConstantSpeed = false; // this messes up downward slope velocity if set to true
 		FloorStopOnSlope = false;   // players should slide on slopes
+
+		PlayerCommands = new EntityCommand[2]
```
[the remark this answers is not in the document]
Comment (policiu) [reply Scripts/Player/Player.cs:97]: Do you think it's worth the clarity loss? I'm indifferent.

File: Scripts/Commands/EntityCommands.cs (hunk 13 above)
Comment (valkyrienyanko) [reply Scripts/Commands/EntityCommands.cs:6]: I personally would prefer `public` added everywhere

File: Scripts/Player/Player.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (valkyrienyanko) [reply Scripts/Player/Player.cs:97]: Just leave it as is then

File: Scripts/Player/Player.cs (hunk 16 above)
Comment (valkyrienyanko) [reply Scripts/Player/Player.cs:248]: I think your changes are good, I just need more explanation on the params of the function. Specifically what is `number` is for and how is it different from `dampening`?

**Review by valkyrienyanko — APPROVED** [review]

File: Scripts/Utils/Godot Helpers/GTimers.cs (hunk 14 above)
Comment (valkyrienyanko) [reply Scripts/Utils/Godot Helpers/GTimers.cs:12]: Still don't understand but it's whatever, not a big deal.

File: Scripts/Player/Player.cs (hunk 16 above)
Comment (policiu) [reply Scripts/Player/Player.cs:248]: Number is the speed value changes. I'll change it to speed and add some comments.

Comment (valkyrienyanko) [reply Scripts/Player/Player.cs:248]: Ohhhhhh that makes sense, I'm dumb

Comment (valkyrienyanko) [reply Scripts/Player/Player.cs:248]: How else would the function return anything useful lol

File: Scripts/Player/Player.cs (hunk 17 above)
Comment (policiu) [reply Scripts/Player/Player.cs:217]: Added
